### DISCUSSION TITLE
Feature Request: `fold` operator in addition to `reduce`

### DISCUSSION THREAD

**Discussion by norcalli** [duckdb/duckdb#13377] [Ideas]
I was attempting to use `reduce` for a complicated aggregation where I was trying to make a `sum_by` macro.

```
CREATE OR REPLACE MACRO sum_by(key, value) AS
list(map {key: value})
.list_reduce((a, v) ->
map_concat(
a,
map_from_entries([{key: k, value: list_sum(a[k] + v[k])} for k in map_keys(v)])
)
);
```

this works, but it would've been nice to use a type other than map for the individual elements while still folding the results into a map, such as using a list of structs and making a map or a list of tuples.

unrelated but it would also be cool if we could have comprehensions over map types as well instead of doing `map_from_entries([{key: e.key, value: e.value for e in source.map_entries()])`, even as just a syntax transform for the equivalent logic.

**Comment by soerenwolfers** [comment]
could you provide a clear description of the proposed `fold` operator, with concrete examples of in and outputs? according to [Wikipedia](https://en.m.wikipedia.org/wiki/Fold_(higher-order_function))
```
fold ([is] also termed reduce, accumulate, aggregate, compress, or inject)
```

**Comment by soerenwolfers** [comment]
your suggestion for `map_from_entries` has mismatched braces. are you aware this function already exists? what's missing?
